File: SlapMac-Windows/AudioManager.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using NAudio.Wave;

namespace SlapMac
{
    /// <summary>
    /// Manages audio playback of slap sound effects.
    /// </summary>
    sealed class AudioManager : IDisposable
    {
        private readonly List<string> _soundFiles = new();
        private readonly Random _random = new();
        private IWavePlayer? _currentPlayer;
        private AudioFileReader? _currentReader;

        public float Volume { get; set; } = 1.0f;
        public bool Enabled { get; set; } = true;
        public int SoundCount => _soundFiles.Count;

        public AudioManager()
        {
            LoadSounds();
        }

        private void LoadSounds()
        {
            var resourceDir = Path.Combine(AppContext.BaseDirectory, "Resources");
            if (!Directory.Exists(resourceDir))
            {
                // Try relative to exe
                resourceDir = Path.Combine(
                    AppContext.BaseDirectory,
                    "Resources");
            }

            if (Directory.Exists(resourceDir))
            {
                var extensions = new HashSet<string>(StringComparer.OrdinalIgnoreCase)
                    { ".mp3", ".wav", ".aiff", ".m4a", ".wma", ".ogg" };

                foreach (var file in Directory.EnumerateFiles(resourceDir))
                {
                    if (extensions.Contains(Path.GetExtension(file)))
                    {
                        _soundFiles.Add(file);
                    }
                }
            }

            System.Diagnostics.Debug.WriteLine($"[SlapMac] Loaded {_soundFiles.Count} sound(s)");
        }

        public void PlayRandomSound()
        {
            if (!Enabled || _soundFiles.Count == 0) return;

            try
            {
                // Dispose previous playback
                StopCurrent();

                int index = _random.Next(_soundFiles.Count);
                _currentReader = new AudioFileReader(_soundFiles[index])
                {
                    Volume = Math.Clamp(Volume, 0f, 1f)
                };

                _currentPlayer = new WaveOutEvent();
                _currentPlayer.Init(_currentReader);
                _currentPlayer.Play();

                System.Diagnostics.Debug.WriteLine(
                    $"[SlapMac] Playing: {Path.GetFileName(_soundFiles[index])}");
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine($"[SlapMac] Playback error: {ex.Message}");
            }
        }

        public bool AddCustomSound(string filePath)
        {
            try
            {
                // Validate extension
                var ext = Path.GetExtension(filePath).ToLowerInvariant();
                var allowed = new HashSet<string> { ".mp3", ".wav", ".aiff", ".m4a", ".wma", ".ogg" };
                if (!allowed.Contains(ext))
                {
                    System.Diagnostics.Debug.WriteLine($"[SlapMac] Unsupported format: {ext}");
                    return false;
                }

                // Sanitize filename
                var fileName = Path.GetFileName(filePath);
                fileName = string.Join("_",
                    fileName.Split(Path.GetInvalidFileNameChars(), StringSplitOptions.RemoveEmptyEntries));

                if (string.IsNullOrWhiteSpace(fileName)) return false;

                var appData = Path.Combine(
                    Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
                    "SlapMac", "Sounds");
                Directory.CreateDirectory(appData);

                var dest = Path.Combine(appData, fileName);
                File.Copy(filePath, dest, overwrite: true);
                _soundFiles.Add(dest);

                System.Diagnostics.Debug.WriteLine($"[SlapMac] Added: {fileName}");
                return true;
            }
            catch (Exception ex)
            {
                System.Diagnostics.Debug.WriteLine($"[SlapMac] Add sound error: {ex.Message}");
                return false;
            }
        }

        private void StopCurrent()
        {
            try
            {
                _currentPlayer?.Stop();
                _currentPlayer?.Dispose();
                _currentReader?.Dispose();
            }
            catch { /* ignore */ }
            _currentPlayer = null;
            _currentReader = null;
        }

        public void Dispose()
        {
            StopCurrent();
        }
    }
}

File: SlapMac-Linux/audio_manager.py
"""Audio playback manager for Linux using pygame."""

import os
import random
import shutil

try:
    import pygame
    pygame.mixer.init(frequency=44100, size=-16, channels=2, buffer=2048)
    HAS_PYGAME = True
except (ImportError, pygame.error):
    HAS_PYGAME = False


class AudioManager:
    def __init__(self, resource_dir):
        self.resource_dir = resource_dir
        self.volume = 1.0
        self.is_enabled = True
        self.sound_files = []
        self._load_sounds()

    def _load_sounds(self):
        if not os.path.isdir(self.resource_dir):
            return

        extensions = {'.mp3', '.wav', '.ogg'}
        for f in sorted(os.listdir(self.resource_dir)):
            ext = os.path.splitext(f)[1].lower()
            if ext in extensions:
                full_path = os.path.join(self.resource_dir, f)
                # Path containment check
                if os.path.abspath(full_path).startswith(
                        os.path.abspath(self.resource_dir)):
                    self.sound_files.append(full_path)

    @property
    def sound_count(self):
        return len(self.sound_files)

    def play_random_sound(self):
        if not self.is_enabled or not self.sound_files or not HAS_PYGAME:
            return

        try:
            path = random.choice(self.sound_files)
            pygame.mixer.music.load(path)
            pygame.mixer.music.set_volume(max(0.0, min(1.0, self.volume)))
            pygame.mixer.music.play()
        except Exception as e:
            print(f"[SlapMac] Playback error: {e}")

    def add_custom_sound(self, filepath):
        if not os.path.isfile(filepath):
            return False

        ext = os.path.splitext(filepath)[1].lower()
        if ext not in {'.mp3', '.wav', '.ogg'}:
            return False

        filename = os.path.basename(filepath)
        filename = filename.replace('..', '').replace('/', '_').replace('\\', '_')
        if not filename or filename.startswith('.'):
            return False

        os.makedirs(self.resource_dir, exist_ok=True)
        dest = os.path.join(self.resource_dir, filename)

        # Path containment check
        if not os.path.abspath(dest).startswith(os.path.abspath(self.resource_dir)):
            return False

        shutil.copy2(filepath, dest)
        self.sound_files.append(dest)
        return True

    def cleanup(self):
        if HAS_PYGAME:
            try:
                pygame.mixer.quit()
            except Exception:
                pass
